src/x32-transposec/rvv.c.in

```diff
@@ -0,0 +1,134 @@
+// Copyright 2023 SiFive, Inc.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+$import math
+$assert SIZE in [32]
+$assert VLEN in [64, 128, 256, 512, 1024]
+$LMUL = 1
+$TILE_WIDTH = int(min(VLEN/SIZE, 8))
+$TILE_HEIGHT = int(VLEN/SIZE)
+$NUM_ITERS = int(math.log2(TILE_HEIGHT))
+
+#include <riscv_vector.h>
+
+#include <assert.h>
+
+#include <xnnpack/common.h>
+#include <xnnpack/math.h>
+#include <xnnpack/transpose.h>
+
+void xnn_x${SIZE}_transposec_ukernel__${TILE_HEIGHT}x${TILE_WIDTH}_rvv(
+    const uint${SIZE}_t* input,
+    uint${SIZE}_t* output,
+    size_t input_stride,
+    size_t output_stride,
+    size_t block_width,
+    size_t block_height,
+    const union xnn_x${SIZE}_transpose_params params[restrict XNN_MIN_ELEMENTS(1)]) XNN_OOB_READS
+{
+  assert(output_stride >= block_height * sizeof(uint${SIZE}_t));
+  assert(input_stride >= block_width * sizeof(uint${SIZE}_t));
+
+  const size_t tile_height = ${TILE_HEIGHT};
+  const size_t tile_width = ${TILE_WIDTH};
+  const size_t tile_hbytes = tile_height * sizeof(uint${SIZE}_t);
+  const size_t tile_wbytes = tile_width * sizeof(uint${SIZE}_t);
+  const size_t input_reset = tile_wbytes - round_down_po2(block_height, tile_height) * input_stride;
+  const size_t input_offset = tile_height * input_stride;
+  const size_t output_reset = tile_width * output_stride - round_down_po2(block_height, 2) * sizeof(uint${SIZE}_t);
+
+  const uint${SIZE}_t* i0 = input;
+
+  uint${SIZE}_t* o0 = (uint${SIZE}_t*) output;
+  $for N in range(1, TILE_WIDTH):
+    uint${SIZE}_t* o${N} = (uint${SIZE}_t*) ((uintptr_t) o${N-1} + output_stride);
+
+  do {
+    size_t bh = block_height;
+    size_t vl = __riscv_vsetvl_e${SIZE}m${LMUL}(tile_height);
+    for (; bh >= ${TILE_HEIGHT}; bh -= ${TILE_HEIGHT}) {
+        if (block_width >= tile_width) {
+            vuint${SIZE}m${LMUL}x${TILE_WIDTH}_t tuple = __riscv_vlsseg${TILE_WIDTH}e${SIZE}_v_u${SIZE}m${LMUL}x${TILE_WIDTH}(i0, input_stride, vl);
```

Comment:
This reads columns of a tile using stride?  That is a slow instruction, no?

Reply:
Based on our benchmark results, this method is much faster than the current one. However, we will continue to try new approaches, and should we prove that one is more effective, we will contribute.